wolfcrypt/src/asn.c

```diff
@@ -13743,190 +13746,99 @@ static int MatchBaseName(int type, const char* name, int nameSz,
-                }
-                base = base->next;
-            } while (base != NULL && !matchDns);
-            /* If we found an DNS type permittedName then name must have had a
-             * match. */
-            if (permittedDns && !matchDns)
-                return 0;
-
-            if (!permittedDns)
+        switch (nameType) {
+            case ASN_DNS_TYPE:
+                /* Should it also consider CN in subject? It could use
+                 * subjectDnsName too */
+                name = cert->altNames;
                 break;
-
-            name = name->next;
-        }
-
-        /* Check each email name matches a permitted. */
-        name = cert->altEmailNames;
-        while (name != NULL) {
-            int matchEmail = 0;
-            int permittedEmail = 0;
-            base = signer->permittedNames;
-            do {
-                /* Looking for permittedNames that are for email. */
-                if (base->type == ASN_RFC822_TYPE) {
-                    permittedEmail = 1;
-                    matchEmail = MatchBaseName(ASN_DNS_TYPE,
-                                               name->name, name->len,
-                                               base->name, base->nameSz);
+            case ASN_RFC822_TYPE:
+                /* Shouldn't it validade E= in subject as well? */
+                name = cert->altEmailNames;
+                break;
+            case ASN_DIR_TYPE:
+                if (cert->subjectRaw != NULL) {
+                    subjectDnsName.next = NULL;
+                    subjectDnsName.type = ASN_DIR_TYPE;
+                    subjectDnsName.len = cert->subjectRawLen;
+                    subjectDnsName.name = (char *)cert->subjectRaw;
+                    name = &subjectDnsName;
                 }
-                base = base->next;
-            } while ((base != NULL) && !matchEmail);
-            /* If we found an email type permittedName then name must have had a
-             * match. */
-            if (permittedEmail && !matchEmail)
-                return 0;
 
-            if (!permittedEmail)
+                #ifndef WOLFSSL_NO_ASN_STRICT
+                /* RFC 5280 section 4.2.1.10
+                    "Restrictions of the form directoryName MUST be
+                    applied to the subject field .... and to any names
+                    of type directoryName in the subjectAltName
+                    extension"
+                */
+                if (name != NULL)
+                    name->next = cert->altDirNames;
+                else
+                    name = cert->altDirNames;
+                #endif
                 break;
-
-            name = name->next;
+            default:
+                /* Other types of names are ignored for now.
+                 * Shouldn't it be rejected if it there is a altNamesByType[nameType]
+                 * and signer->extNameConstraintCrit is set? */
+                return 0;
```

Comment:
Yes I would agree on rejecting if `extNameConstraintCrit`. Can you add this?

Reply:
I will need to add some extra fields as signer does not have extNameConstraintCrit

Reply:
Thinking about this one, wouldn't it be better to implement all supported name types in name constraints? If a name type is unsupported, it would be a certificate parser job to reject it if it is critical. Once the parser reads it, I would expected that everything else should support it as well. It is missing only ASN_URI_TYPE and ASN_IP_TYPE